hat/app/org/hatdex/hat/api/controllers/ContractData.scala

```diff
@@ -0,0 +1,583 @@
+              case (_, Left(x)) => handleFailedRequestAssessment(x)
+            }
+          }
+        case None => Future.successful(BadRequest("Missing Contract Details."))
+      }
+    }
+
+  def createContractData(
+      namespace: String,
+      endpoint: String,
+      skipErrors: Option[Boolean]): Action[ContractDataCreateRequest] =
+    UserAwareAction.async(parsers.json[ContractDataCreateRequest]) { implicit request =>
+      val contractDataCreate    = request.body
+      val maybeContractDataInfo = refineContractDataCreateRequest(contractDataCreate)
+      maybeContractDataInfo match {
+        case Some(contractDataInfo) =>
+          contractValid(contractDataInfo, namespace).flatMap { contractOk =>
+            contractOk match {
+              case (Some(hatUser), Right(RequestVerified(ns))) =>
+                handleCreateContractData(hatUser, contractDataCreate, ns, endpoint, skipErrors)
+              case (_, Left(x)) => handleFailedRequestAssessment(x)
+            }
+          }
+        case None => Future.successful(BadRequest("Missing Contract Details."))
+      }
+
+      Future.successful(BadRequest("Not Implemented."))
+    }
+
+  def updateContractData(namespace: String): Action[ContractDataUpdateRequest] =
+    UserAwareAction.async(parsers.json[ContractDataUpdateRequest]) { implicit request =>
+      val contractDataUpdate    = request.body
+      val maybeContractDataInfo = refineContractDataUpdateRequest(contractDataUpdate)
+      maybeContractDataInfo match {
+        case Some(contractDataInfo) =>
+          contractValid(contractDataInfo, namespace).flatMap { contractOk =>
+            contractOk match {
+              case (Some(hatUser), Right(RequestVerified(ns))) =>
+                handleUpdateContractData(hatUser, contractDataUpdate, ns)
+              case (_, Left(x)) => handleFailedRequestAssessment(x)
+              case (_, _)       => handleFailedRequestAssessment(GeneralError)
+            }
+          }
+        case None => Future.successful(BadRequest("Missing Contract Details."))
+      }
+
+      Future.successful(BadRequest("Not Implemented."))
+    }
+
+  // -- Pull this out from RichData and ContractData - this is duplicated --
+  private def makeData(
+      namespace: String,
+      endpoint: String,
+      orderBy: Option[String],
+      ordering: Option[String],
+      skip: Option[Int],
+      take: Option[Int]
+    )(implicit db: HATPostgresProfile.api.Database): Future[Result] = {
+    val dataEndpoint = s"$namespace/$endpoint"
+    val query =
+      Seq(EndpointQuery(dataEndpoint, None, None, None))
+    val data = dataService.propertyData(
+      query,
+      orderBy,
+      ordering.contains("descending"),
+      skip.getOrElse(0),
+      take.orElse(Some(defaultRecordLimit))
+    )
+    data.map(d => Ok(Json.toJson(d)))
+  }
+
+  trait ContractRequest {
+    def refineContract: ContractDataInfoRefined
+  }
+
+  def refineContractInfo(
```

Comment:
Non Blocking: Out of scope for this PR but Validated [https://typelevel.org/cats/datatypes/validated.html] would be realy useful here as it would vaildate all fields and return a list of errors instead of swallowing the error

Reply:
I'll look into Validated, maybe it's a quick add.

Reply:
Its basically an async either and it really nice for this purpose, @xavji has been doing some great work with it for validating in dev portal if you want some examples from out codebase. Maybe not a quick add but something to keep in mind for the future